### PR TITLE
Generic interceptors

### PR DESCRIPTION
**Pull request by RikkiGibson**
Test plan: #67421 
Continuation of #68218 

Addresses point 1 in https://github.com/dotnet/roslyn/pull/68218#issuecomment-1585161986

> We want to deliver a critical level of generics support in .NET 8. Namely, in addition to permitting interceptors with arity 0, we will also permit interceptors whose arity equals the _net arity_ of the original method (sum of arity of containing types plus arity of original method). The type arguments from the original method and containing types are passed along to the interceptor in source order, i.e. starting from the outermost generic type, and inwards to the original method.

It might also be worthwhile to spend a little more time investigating the "unification" approach. But it's something that could be added later without breaking things.

### CODE REVIEW

File: docs/features/interceptors.md [hunk 1]
```diff
@@ -127,7 +149,7 @@ static class Program
 static class D
 {
     [InterceptsLocation("Program.cs", 12, 11)]
-    public static void Interceptor1(object s) => throw null!;
+    public static void Interceptor1<T1>(T2 t) => throw null!;
```
Comment (cston) [inline docs/features/interceptors.md:152]: >T1>(T2 [](http://example.com/codeflow?start=36&length=6)

Should these two names be the same?

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 2]
```diff
@@ -148,7 +148,46 @@ private void EmbedIfNeedTo(BoundExpression receiver, ImmutableArray<PropertySymb
             }
 
             Debug.Assert(interceptableLocation != null);
-            Debug.Assert(interceptor.Arity == 0);
+            Debug.Assert((object)interceptor.OriginalDefinition == interceptor);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:151]: >(object)interceptor.OriginalDefinition == interceptor [](http://example.com/codeflow?start=25&length=53)

Minor: `interceptor.IsDefinition`

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 3]
```diff
@@ -148,7 +148,46 @@ private void EmbedIfNeedTo(BoundExpression receiver, ImmutableArray<PropertySymb
...
+                    stack.Push(type);
+                }
+
+                var typeArgumentsBuilder = ArrayBuilder<TypeSymbol>.GetInstance();
+                do
+                {
+                    // gather the type arguments starting from outermost containing type in to the method.
+                    typeArgumentsBuilder.AddRange(stack.Pop().GetMemberTypeArgumentsNoUseSiteDiagnostics());
+                }
+                while (stack.Count != 0);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:170]: Consider using:
```csharp
var builder = ArrayBuilder<TypeWithAnnotations>.GetInstance();
method.ContainingType.GetAllTypeArgumentsNoUseSiteDiagnostics(builder);
builder.AddRange(method.TypeArgumentsWithAnnotations);
interceptor = interceptor.Construct(builder.ToImmutableAndFree());
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs [hunk 4]
```diff
@@ -1799,6 +1797,393 @@ static class D
...
+                [InterceptsLocation("Program.cs", 17, 19)]
+                public static void Interceptor1<T>(T t) { Console.Write(t); }
+            }
+            """;
+        var verifier = CompileAndVerify(new[] { (source, "Program.cs"), s_attributesSource }, parseOptions: RegularWithInterceptors, expectedOutput: "1122");
+        verifier.VerifyDiagnostics();
+    }
+
+    [Fact]
+    public void InterceptableGeneric_08()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs:1898]: >InterceptableGeneric_08 [](http://example.com/codeflow?start=16&length=23)

Is this distinct from `_07`?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs:1898]: It's not, sorry. Deleting.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 5]
```diff
@@ -148,7 +148,43 @@ private void EmbedIfNeedTo(BoundExpression receiver, ImmutableArray<PropertySymb
...
 
             Debug.Assert(interceptableLocation != null);
-            Debug.Assert(interceptor.Arity == 0);
+            Debug.Assert(interceptor.IsDefinition);
+            Debug.Assert(!interceptor.ContainingType.IsGenericType);
+
+            if (interceptor.Arity != 0)
+            {
+                // Gather the type arguments of the original method and its containing types, from outermost to innermost.
+                var stack = ArrayBuilder<Symbol>.GetInstance();
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:157]: It looks like this stack doesn't get used

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 6]
```diff
@@ -148,7 +148,43 @@ private void EmbedIfNeedTo(BoundExpression receiver, ImmutableArray<PropertySymb
...
+                }
+                else if (interceptor.Arity != netArity)
+                {
+                    this._diagnostics.Add(ErrorCode.ERR_InterceptorArityNotCompatible, attributeLocation, interceptor, netArity, method);
+                    typeArgumentsBuilder.Free();
+                    return;
+                }
+
+                interceptor = interceptor.Construct(typeArgumentsBuilder.ToImmutableAndFree());
+                if (!interceptor.CheckConstraints(new ConstraintsHelper.CheckConstraintsArgs(this._compilation, this._compilation.Conversions, includeNullability: false, attributeLocation, this._diagnostics)))
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:183]: >includeNullability: false [](http://example.com/codeflow?start=143&length=25)

Why not include nullability?

File: src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs [hunk 7]
```diff
@@ -1799,6 +1797,360 @@ static class D
...
+                public static void Usage()
+                {
+                    C.InterceptableMethod("abc");
+                }
+            }
+
+            static class D
+            {
+                [InterceptsLocation("Program.cs", 10, 11)] // 1
+                public static void Interceptor1<T>(T t) where T : struct => throw null!;
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs:1854]: >where T : struct [](http://example.com/codeflow?start=56&length=16)

Let's also test with constraints relative to nullability (a nullable or non-nullable reference type, or the `notnull` constraint)

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs [hunk 8]
```diff
@@ -148,7 +148,43 @@ private void EmbedIfNeedTo(BoundExpression receiver, ImmutableArray<PropertySymb
...
+            {
+                // Gather the type arguments of the original method and its containing types, from outermost to innermost.
+                var stack = ArrayBuilder<Symbol>.GetInstance();
+                stack.Push(method);
+                for (var type = method.ContainingType; type is not null; type = type.ContainingType)
+                {
+                    stack.Push(type);
+                }
+
+                var typeArgumentsBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance();
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs:164]:
```suggestion
                var typeArgumentsBuilder = ArrayBuilder<TypeWithAnnotations>.GetInstance();
```

File: src/EditorFeatures/CSharp/LanguageServer/CSharpLspBuildOnlyDiagnostics.cs [hunk 9]
```diff
@@ -55,6 +55,8 @@ namespace Microsoft.CodeAnalysis.CSharp.LanguageServer
         "CS9159", // ErrorCode.WRN_NullabilityMismatchInParameterTypeOnInterceptor
         "CS9160", // ErrorCode.ERR_InterceptorCannotInterceptNameof
         "CS9163"  // ErrorCode.ERR_SymbolDefinedInAssembly
```
Comment (jcouv) [inline src/EditorFeatures/CSharp/LanguageServer/CSharpLspBuildOnlyDiagnostics.cs:57]: Looks like a comma is missing here

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7) modulo missing comma

**Pull request merged**